scala/private/common.bzl

```diff
@@ -0,0 +1,174 @@
+# causing a conflict between objects and packages warning
+#  error: package cats contains object and package with same name: implicits
+# one of them needs to be removed from classpath
+# import cats.implicits._
+
+def not_sources_jar(name):
+  return "-sources.jar" not in name
+
+def filter_not_sources(deps):
+  return depset([dep for dep in deps.to_list() if not_sources_jar(dep.basename) ])
+
+def add_labels_of_jars_to(jars2labels, dependency, all_jars, direct_jars):
+  for jar in direct_jars:
+      _add_label_of_direct_jar_to(jars2labels, dependency, jar)
+  for jar in all_jars:
+      _add_label_of_indirect_jar_to(jars2labels, dependency, jar)
+
+def _add_label_of_direct_jar_to(jars2labels, dependency, jar):
+    jars2labels[jar.path] = dependency.label
+
+def _add_label_of_indirect_jar_to(jars2labels, dependency, jar):
+ if _label_already_exists(jars2labels, jar):
+   return
+
+ # skylark exposes only labels of direct dependencies.
+ # to get labels of indirect dependencies we collect them from the providers transitively
+ if _provider_of_dependency_contains_label_of(dependency, jar):
+     jars2labels[jar.path] = dependency.jars_to_labels[jar.path]
+ else:
+     jars2labels[jar.path] = "Unknown label of file {jar_path} which came from {dependency_label}".format(
+         jar_path = jar.path,
+         dependency_label = dependency.label
+     )
+
+def _label_already_exists(jars2labels, jar):
+  return jar.path in jars2labels
+
+def _provider_of_dependency_contains_label_of(dependency, jar):
+  return hasattr(dependency, "jars_to_labels") and jar.path in dependency.jars_to_labels
+
+def create_java_provider(scalaattr, transitive_compile_time_jars):
+    # This is needed because Bazel >=0.7.0 requires ctx.actions and a Java
+    # toolchain. Fortunately, the same change that added this requirement also
+    # added this field to the Java provider so we can use it to test which
+    # Bazel version we are running under.
+    test_provider = java_common.create_provider()
+
+    if hasattr(test_provider, "full_compile_jars"):
+      return java_common.create_provider(
+          use_ijar = False,
+          compile_time_jars = scalaattr.compile_jars,
+          runtime_jars = scalaattr.transitive_runtime_jars,
+          transitive_compile_time_jars = depset(transitive = [transitive_compile_time_jars, scalaattr.compile_jars]),
+          transitive_runtime_jars = scalaattr.transitive_runtime_jars,
+      )
+    else:
+      return java_common.create_provider(
+          compile_time_jars = scalaattr.compile_jars,
+          runtime_jars = scalaattr.transitive_runtime_jars,
+          transitive_compile_time_jars = transitive_compile_time_jars,
+          transitive_runtime_jars = scalaattr.transitive_runtime_jars,
+      )
+
+# TODO: this should really be a bazel provider, but we are using old-style rule outputs
+# we need to document better what the intellij dependencies on this code actually are
+def create_scala_provider(
```

Comment:
I don't think this should be private. We eventually want to have an API to create a ScalaInfo. Can we move to public?

Reply:
Just `create_scala_provider`? or `create_scala_provider` and `create_java_provider`?

Reply:
create_java_... is a compatability tool for our use, so I would say it is private, other callers should use the bazel API. So I would just make create_scala_provider the public thing.